src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,87 @@
+/**
+ * CS2103T Individual Project.
+ *
+ * @author Wang Jiefan
+ * @version 1.0
+ * @since 1/20/2023
+ */
+
+package duke;
+
+import duke.command.ArgCommand;
+import duke.command.BasicCommand;
+import duke.command.Command;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+/**
+ * The main class of the Duke program.
+ */
+public class Duke {
+
+  private final Ui ui;
+  private final Parser parser;
+
+  private Duke(String filename) {
+    this.ui = new Ui();
+    TaskList taskList;
+    try {
+      taskList = new TaskList(filename);
+    } catch (Exception e) {
+      ui.loadError(e);
+      taskList = new TaskList(new ArrayList<>());
+    }
+    Command[] commands = new Command[]{
+        new BasicCommand("exit", "exit the app", () -> new String[]{"Goodbye."}),
+        new BasicCommand("help", "show this help message", ui::getHelpMsg),
+        new BasicCommand("list", "list tasks", taskList::stringify),
+        new ArgCommand("add", "add task", new String[]{"\\s"}, taskList::add),
+        new ArgCommand("mark", "mark/unmark task as done", new String[]{}, taskList::mark),
+        new ArgCommand("delete", "delete task", new String[]{}, taskList::delete),
+        new ArgCommand("find", "find tasks containing text fragment",
+            new String[]{}, taskList::find)};
```

Comment:
I like how you have used functional programming approach for this project while at the same time using OO 😮